scylla/src/transport/session.rs

```diff
@@ -599,6 +599,19 @@ impl Session {
                 .extend(statement.prepare_tracing_ids);
         }
 
+        let keyspace = prepared.get_keyspace_name();
+        let table = prepared.get_table_name();
+        if keyspace.is_some() && table.is_some() {
```

Comment:
This:
```
fn extract_partitioner_name(&self, prepared: &PreparedStatement) -> Option<&str> {
        let cluster_data = self.cluster.get_data();
        cluster_data
            .keyspaces
            .get(prepared.get_keyspace_name()?)?
            .tables
            .get(prepared.get_table_name()?)?
            .partitioner
    }
 ```
does not work - the types do not match (`Option<&str>` vs Option<String>). Is it even possible to cast it? Any usual tricks such as `.as_deref()` end with result referencing data owned by this method.

Reply:
This works:

```rust
fn extract_partitioner_name<'a>(
    &self,
    prepared: &PreparedStatement,
    cluster_data: &'a ClusterData,
) -> Option<&'a str> {
    cluster_data
        .keyspaces
        .get(prepared.get_keyspace_name()?)?
        .tables
        .get(prepared.get_table_name()?)?
        .partitioner
        .as_ref()
        .map(|s| s.as_ref())
}
```

You need to get inside the optional in order to get a `&str` from `String`, hence the `.map`.

Note that the returned string borrows from the `cluster_data`. You need to provide it from outside and give explicit lifetimes in order for everything to check out, otherwise Rust will rightfully complain that you are returning a reference to an object that might potentially no longer exist.